### ISSUE TITLE
python-casacore >= 2.1.2 to python-casacore == 2.1.2

### ISSUE THREAD

**Issue by radiosky-abhik** [ratt-ru/CubiCal#92]
Hi Jon,
            Cubical installation in our Rhodes cluster is giving some python casacore issues. Looking at the error Ben suggested python-casacore == 2.1.2 instead of python-casacore >= 2.1.2 and it works. 

PS. Now it only complains about "Failed building wheel for vext.pyqt4"

- Abhik

**Comment by o-smirnov** [comment]
We need to ask @gijzelaerr to look at this properly. I don't like fixed version number dependencies. There's nothing it CubiCal that **requires** 2.1.2 specifically, it's more likely some mix-up between casacore and python-casacore that happens during installation.

**Comment by JSKenyon** [comment]
This is likely a result of the new python-casacore release. So the latest python-casacore depends on a newer version casa and thus fails if it installs the casacore from the radio-astro or KERN2 PPAs. Though I am no expert on this stuff. 

The failed wheel build for vext is normal and can safely be ignored.

**Comment by gijzelaerr** [comment]
i've uploaded python-casacore 2.2.0 to KERN-dev and KERN-3 (not released yet).

**Comment by JSKenyon** [comment]
Just drawing attention back to this @o-smirnov and @gijzelaerr. What is the best way to deal with this going forward? I can of course change the requirement, as suggested by Abhik, but this seems brittle and dangerous. The fact that the current pipable python-casacore depends on a debian package in KERN-3 is very frustrating and will make backwards compatibility tough.

**Comment by o-smirnov** [comment]
Honestly I don't know. I have to defer this to @gijzelaerr.  I suppose this is the crux of the problem is this:

> The fact that the current pipable python-casacore depends on a debian package in KERN-3

``casacore`` and ``python-casacore`` need to be installed in lockstep, so if we rely on pip for one and KERN for the other, this will arise with every casacore release.

**Comment by JSKenyon** [comment]
Ok, for now I am going to lock the python-casacore version to 2.1.2. I will leave this issue open, but I do not want new users to have to struggle through installation right now.

**Comment by JSKenyon** [comment]
@o-smirnov Basic fix in #98. Please ok it.

**Comment by o-smirnov** [comment]
Not convinced locking to 2.1.2 is a solution, because that makes it incompatible with KERN-3/KERN-dev.

> I do not want new users to have to struggle

Fair enough, but first consider, where are your new users installing this? The Rhodes machines are (as of recently) tracking KERN-3/dev, so now you've broken it for Rhodes users. Which I reckon is a major chunk of the new user demographic at the moment. :)

There's no way to make everyone happy here, because fundamentally there's no foolproof way to pip install python-casacore (unless it is made to bring in its own casacore, @gijzelaerr?). Maybe it's better to recommend that the users ask their admins for a working systemwide python-casacore (which is usually already the case in the first place), and make the virtualenv with --system-site-packages?

**Comment by gijzelaerr** [comment]
if you use KERN-3 or KERN-dev then you can just use the python-casacore debian package.

meanwhile, i've made a new python-casacore release that makes it pip installable again:

https://github.com/casacore/python-casacore/releases/tag/v2.2.1

If travis is happy it will upload this to pypi.

**Comment by JSKenyon** [comment]
@o-smirnov I really like not requiring --system-site-packages. But I see your point. Guess I will think about this a little more and wait for @gijzelaerr to return.

**Comment by gijzelaerr** [comment]
https://pypi.python.org/pypi/python-casacore there we go, enjoy.

Greeting from Bahrain, king Hamad bin Isa Al Khalifa says hi.

**Comment by o-smirnov** [comment]
> Greeting from Bahrain, king Hamad bin Isa Al Khalifa says hi.

Thanks, enjoy Bahrain. I'm still not quite clear on the situation. OK, so you put a new python-casacore on pip that's compatible with the casacore in KERN-3/dev -- doesn't this now break things for KERN-2 users? I only see three robust ways out of this:

* make the setup.py script intelligent, so it figures out which casacore is on the system, and pip installs the appropriate python-casacore (@gijzelaerr will hate this)

* rely on --system-site-packages (@JSKenyon will hate this. I'm ok with it -- any sensible system will already have a python-casacore one way or another)

* distribute casacore itself via pip (everyone will hate this).

**Comment by gijzelaerr** [comment]
eeeeeh what do you think the problem actually is here. python-casacore is a KERN package also, use that if you want stability, if you need a newer version do a pip install. There is no problem here if you ask me.

**Comment by o-smirnov** [comment]
> python-casacore is a KERN package also, use that if you want stability.

Which means creating the virtualenv with ``--system-site-packages``. That's fine with me. @JSKenyon, how about we put this in the instructions:

* if you have a working python-casacore on your system, please create the virtualenv with ``--system-site-packages``

* if you don't, don't -- but please pip install the correct version of python-casacore (correct means matching your casacore binary) into your virtual environment before you install CubiCal

**Comment by JSKenyon** [comment]
Ok - that is probably the most sensible. Will update the PR and docs accordingly.

**Comment by gijzelaerr** [comment]
so not that i'm against adding this kind of sections to your doc, i just think it will probably break again at some point. Maybe better to write a central manual on how to use KERN and/or pip and/or casacore and not try to copy paste this into every project.

**Comment by JSKenyon** [comment]
OK - I am going to update the revised setup pull request. I think I am going to deal with this by removing the python-casacore from the setup.py and instead make a prerequisite that you install BEFORE installing CubiCal. This should make it clear to users when the installation of python-casacore is failing. It will also be easy to have a few lines in the installation docs explaining which python-casacore should be installed based on the version of casacore. I think this should be relatively simple, avoids --system-site-packages, and also makes it clear when the installation failure is outside of CubiCal. Thoughts?

**Comment by gijzelaerr** [comment]
I don't understand the problem and why you need to do this. What is actually the problem with the current KERN package and/or the pip version of python-casacore?

**Comment by JSKenyon** [comment]
We recommend installing CubiCal in a virtualenv. This just makes things cleaner. However, unless you create the virtualenv using --system-site-packages, python packages installed outside the virtualenv are not visible. To get around this, we usually pip install python-casacore into the virtualenv (see [the docs](http://cubical.readthedocs.io/en/latest/installation.html)). By default, this grabs the latest python-casacore, which may not be compatible with the user's installed casacore. It should all work fine if you use kern-dev, and I am currently checking that. However, for people still on 14.04 (such as myself) it is necessary to use an older python-casacore. This means the user needs to install a compatible python-casacore before installing CubiCal. I guess the other alternative to just abandon 14.04 support.

**Comment by gijzelaerr** [comment]
you can solve this by:
* removing the specific python-casacore version dependecy (but still depend on python-casacore)
* install an older python-casacore from pip before you install cubical.

And probably you should just update your system.

**Comment by JSKenyon** [comment]
This isn't a real problem for me - I know how to get it working. Just trying to figure out the easiest option for end users who may not be familiar with all the pieces. But I suspect what I will do is make the 16.04 instructions use the kern-dev PPA and then add a small section on how to make it work on 14.04. Thanks for all the help.

**Comment by o-smirnov** [comment]
> And probably you should just update your system.

Yes, **we** don't have a problem at all. @JSKenyon just wants to give the clearest possible installation instructions for naive users. But I think simply requiring them to have a working python-casacore is not too onerous.

**Comment by JSKenyon** [comment]
I am struggling to get a docker container working with KERN-3. My Dockerfile looks like this.

```
# Use an Ubuntu as parent image.
FROM ubuntu:16.04

# Set the working directory to /docker_test
WORKDIR /docker_test

# Copy the current directory contents into the container at /docker_test
ADD . /docker_test

RUN apt-get update
RUN apt-get install -y git
RUN apt-get install -y python-pip
RUN apt-get install -y libboost-all-dev

RUN apt-get install -y software-properties-common
RUN apt-add-repository -s ppa:kernsuite/kern-3 # For 16.04
RUN apt-add-repository multiverse
RUN apt-add-repository restricted
RUN apt-get update
RUN apt-get install -y python-casacore

RUN pip install -U pip setuptools wheel virtualenv
RUN virtualenv cubienv
RUN . cubienv/bin/activate && \
	pip install -U pip setuptools wheel && \
	pip install numpy && \
	pip install git+https://github.com/ratt-ru/CubiCal.git@rev-setup
```
I would expect python-casacore (installed via the PPA) would provide a casacore which is compatible with the current pipable python-casacore. Instead I get this warning:

```
UserWarning: Your casacore version is older than 2.3.0 and incompatible with this version of python-casacore
```

The installation then fails with:
```
src/fit.cc:31:53: fatal error: casacore/python/Converters/PycBasicData.h: No such file or directory
```
 It is possible that I am being an idiot, but I am struggling to figure out why. Is it possibly just something path related?

**Comment by gijzelaerr** [comment]
Ok i have several comments here:
* Use the kern docker image
* use the docker-apt-install command bundled with this image
* dont do a git checkout of your project but bundle your project with the dockerfile
* use debian packages where possible
* you don't need to build a virtualenv in a docker container, it self is already an isolated environment.


try this (might contain a typo), put this dockerfile in the root of your project and build it:

```
FROM kernsuite/base:3
RUN docker-apt-install casacore-dev python-casacore python-pip python-numpy libboost-all-dev
ADD . /code
WORKDIR /code
RUN pip install .
```

**Comment by JSKenyon** [comment]
Except, and forgive my frustration, that doesn't actually accomplish what I am trying to do. I am not trying to build a docker image for deployment - I am trying to create a fictitious installation environment which will replicate the standard CubiCal installation process on a 16.04 machine using KERN-3. I know the virtualenv is not necessary when using docker - but it would be for a user on 16.04. Point is, this should work, doesn't and I would like to know why.

**Comment by JSKenyon** [comment]
For reference, this works:

```
# Use an Ubuntu as parent image.
FROM ubuntu:16.04

# Set the working directory to /docker_test
WORKDIR /docker_test

# Copy the current directory contents into the container at /docker_test
ADD . /docker_test

RUN apt-get update
RUN apt-get install -y git
RUN apt-get install -y python-pip
RUN apt-get install -y libboost-all-dev

RUN apt-get install -y software-properties-common
RUN apt-add-repository -s ppa:kernsuite/kern-3 # For 16.04
RUN apt-add-repository multiverse
RUN apt-add-repository restricted
RUN apt-get update
RUN apt-get install -y python-casacore

RUN pip install -U pip setuptools wheel virtualenv
RUN pip install -U numpy
RUN pip install git+https://github.com/ratt-ru/CubiCal.git@rev-setup
```

**Comment by gijzelaerr** [comment]
You are violating my 4th commandment.

If you don't use the KERN package but pip install python-casacore you need to have casacore-dev installed.
```
gijs@jake:~
λ   dpkg -S casacore/python/Converters/PycBasicData.h
casacore-dev: /usr/include/casacore/python/Converters/PycBasicData.h
```

**Comment by gijzelaerr** [comment]
ah wait, i was to quick. so you install the debian package for python-casacore, but then install in a clean virtualenv. This will compile python-casacore, which requires casacore-dev to be installed.

I'm not sure what that warning from python-casacore means, that is just wrong.

**Comment by gijzelaerr** [comment]
this is documented here:

https://github.com/casacore/python-casacore#using-pip

```
apt-get install casacore-dev python-numpy \
    python-setuptools libboost-python-dev libcfitsio3-dev wcslib-dev
```

**Comment by gijzelaerr** [comment]
btw, if that cpython issue is fixed i'll make a package for KERN-3, which would greatly simplify your installation instructions.

**Comment by sjperkins** [comment]
@gijzelaerr python-casacore doesn't appear to install the most general soft link of the .so files. For example:
```
root@ab25c45b9abb:/docker_test# ls -l /usr/lib/x86_64-linux-gnu/libcasa_*
lrwxrwxrwx 1 root root      21 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_casa.so.2 -> libcasa_casa.so.2.4.0
-rw-r--r-- 1 root root 3768696 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_casa.so.2.4.0
lrwxrwxrwx 1 root root      28 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_coordinates.so.2 -> libcasa_coordinates.so.2.4.0
-rw-r--r-- 1 root root 1050968 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_coordinates.so.2.4.0
lrwxrwxrwx 1 root root      21 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_fits.so.2 -> libcasa_fits.so.2.4.0
-rw-r--r-- 1 root root 1349672 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_fits.so.2.4.0
lrwxrwxrwx 1 root root      23 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_images.so.2 -> libcasa_images.so.2.4.0
-rw-r--r-- 1 root root 3156256 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_images.so.2.4.0
lrwxrwxrwx 1 root root      25 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_lattices.so.2 -> libcasa_lattices.so.2.4.0
-rw-r--r-- 1 root root 2846792 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_lattices.so.2.4.0
lrwxrwxrwx 1 root root      25 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_measures.so.2 -> libcasa_measures.so.2.4.0
-rw-r--r-- 1 root root 1877232 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_measures.so.2.4.0
lrwxrwxrwx 1 root root      23 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_mirlib.so.2 -> libcasa_mirlib.so.2.4.0
-rw-r--r-- 1 root root  137664 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_mirlib.so.2.4.0
lrwxrwxrwx 1 root root      19 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_ms.so.2 -> libcasa_ms.so.2.4.0
-rw-r--r-- 1 root root 4539064 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_ms.so.2.4.0
lrwxrwxrwx 1 root root      23 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_python.so.2 -> libcasa_python.so.2.4.0
-rw-r--r-- 1 root root  233880 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_python.so.2.4.0
lrwxrwxrwx 1 root root      24 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_scimath.so.2 -> libcasa_scimath.so.2.4.0
-rw-r--r-- 1 root root 1448024 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_scimath.so.2.4.0
lrwxrwxrwx 1 root root      26 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_scimath_f.so.2 -> libcasa_scimath_f.so.2.4.0
-rw-r--r-- 1 root root  449072 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_scimath_f.so.2.4.0
lrwxrwxrwx 1 root root      23 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_tables.so.2 -> libcasa_tables.so.2.4.0
-rw-r--r-- 1 root root 7046600 Nov  6 14:35 /usr/lib/x86_64-linux-gnu/libcasa_tables.so.2.4.0
```

As you say, `casacore-dev` must also be installed to get the most general .so soft link.

**Comment by gijzelaerr** [comment]
Python-casaore doesn't install anything in /usr/lib. It depends on various shared libraries that get installed when you install that package. the -dev package installs the generic .so links, which are used during linking, not runtime. That is how all debian packages are packaged. To link against a library you need to install the -dev package.

**Comment by sjperkins** [comment]
OK, https://lists.debian.org/debian-mentors/2006/03/msg00356.html confirms that.

**Comment by JSKenyon** [comment]
Ok - finally getting a handle on this. I can require the installation of those dependencies, then install python-casacore into the virtualenv. This does work inside my docker container - should work for most users. For 14.04, the only change should be requiring users to install a python-casacore version compatible with their installed casacore. Should be 2.1.2. I am just testing this out, then will update my revised setup PR.

**Comment by JSKenyon** [comment]
By the look of it casacore-dev isn't available into radio-astro. Is that the case?

**Comment by JSKenyon** [comment]
Never mind - sorted it out with @sjperkins help. Thanks everyone who pitched in here.

**Comment by JSKenyon** [comment]
Once #89 is accepted and merged, this problem should be resolved.

**Comment by JSKenyon** [comment]
Fixed in #89
